src/Functions/timeSlots.cpp

```diff
@@ -129,67 +224,150 @@ class FunctionTimeSlots : public IFunction
+                if (!time_slot_column)
+                    throw Exception("Third argument for function " + getName() + " must be constant UInt32", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
 
-        auto res = ColumnArray::create(ColumnUInt32::create());
-        ColumnUInt32::Container & res_values = typeid_cast<ColumnUInt32 &>(res->getData()).getData();
+                if (time_slot_size = time_slot_column->getValue<UInt32>(); time_slot_size == 0)
+                    throw Exception("Third argument for function " + getName() + " must be greater than zero", ErrorCodes::ILLEGAL_COLUMN);
+            }
 
-        auto time_slot_size = TIME_SLOT_SIZE;
+            const auto * dt_starts = checkAndGetColumn<ColumnUInt32>(arguments[0].column.get());
+            const auto * dt_const_starts = checkAndGetColumnConst<ColumnUInt32>(arguments[0].column.get());
 
-        if (arguments.size() == 3)
-        {
-            const auto * time_slot_column = checkAndGetColumn<ColumnConst>(arguments[2].column.get());
-            if (!time_slot_column)
-                throw Exception("Third argument for function " + getName() + " must be constant UInt32", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            const auto * durations = checkAndGetColumn<ColumnUInt32>(arguments[1].column.get());
+            const auto * const_durations = checkAndGetColumnConst<ColumnUInt32>(arguments[1].column.get());
 
-            if (time_slot_size = time_slot_column->getValue<UInt32>(); time_slot_size == 0)
-                throw Exception("Third argument for function " + getName() + " must be greater than zero", ErrorCodes::ILLEGAL_COLUMN);
-        }
+            auto res = ColumnArray::create(ColumnUInt32::create());
+            ColumnUInt32::Container & res_values = typeid_cast<ColumnUInt32 &>(res->getData()).getData();
 
-        if (starts && durations)
-        {
-            TimeSlotsImpl<UInt32>::vectorVector(starts->getData(), durations->getData(), time_slot_size, res_values, res->getOffsets());
-            return res;
+            if (dt_starts && durations)
+            {
+                TimeSlotsImpl::vectorVector(dt_starts->getData(), durations->getData(), time_slot_size, res_values, res->getOffsets());
+                return res;
+            }
+            else if (dt_starts && const_durations)
+            {
+                TimeSlotsImpl::vectorConstant(dt_starts->getData(), const_durations->getValue<UInt32>(), time_slot_size, res_values, res->getOffsets());
+                return res;
+            }
+            else if (dt_const_starts && durations)
+            {
+                TimeSlotsImpl::constantVector(dt_const_starts->getValue<UInt32>(), durations->getData(), time_slot_size, res_values, res->getOffsets());
+                return res;
+            }
         }
-        else if (starts && const_durations)
+        else if (isDateTime64(arguments[0].type))
```

Comment:
No need to check that arguments[0].type is a DateTime64 - this is already ensured by getReturnTypeImpl(). Perhaps you could replace l. 311 by an "else" + add an "assert(isDateTime64(arguments[0].type))".